File: longest valid paranthesis.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        l=len(s)
        st=[]
        t=[0]*l
        m=0
        su=0
        for i in range(l):
            if s[i]=='(':
                st+=[i]
            else:
                if len(st):
                    t[st[-1]]=1
                    t[i]=1
                    st.pop()
        for j in range(l):
            if t[j]:
                su+=t[j]
                if j==l-1:
                    if su>m:
                        m=su
            else:
                if su>m:
                    m=su
                su=0
        print(t)
        return m

File: longest valid paranthesis.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int l=s.size();
        stack<int>st;
        vector<int>t(l,0);
        int m=0;
        int su=0;
        for (int i=0;i<l;i++)
        {
            if(s[i]=='(')
                st.push(i);
            else
                if(st.size())
                {
                    t[st.top()]=1;
                    t[i]=1;
                    st.pop();
                }
        }
        for(int j=0;j<l;j++)
        {
            if (t[j])
            {
                su+=t[j];
                if(j==l-1)
                    if(su>m)
                        m=su;
            }
            else
            {
                if(su>m)
                    m=su;
                su=0;
            }
        }
        return m; 
    }
};
